File: src/connections/apify_connection.py
#!/usr/bin/env python3
"""
Apify API Connection Class
Handles direct communication with the Apify API for web scraping tasks
"""
import os
import logging
import json
from typing import Dict, List, Optional, Any, Union

import httpx
from dotenv import load_dotenv
from apify_client import ApifyClient

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s",
)
logger = logging.getLogger(__name__)

class ApifyConnection:
    """Connection class for Apify API integration"""
    
    def __init__(self, api_token: Optional[str] = None):
        """Initialize Apify connection with API token"""
        # Load environment variables if token not provided
        if not api_token:
            load_dotenv()
            api_token = os.environ.get("APIFY_API_TOKEN")
        
        self.api_token = api_token
        if not self.api_token:
            raise ValueError("Apify API token not provided")
        
        # Initialize HTTP client for direct API calls
        self.client = httpx.AsyncClient(timeout=120.0)  # 2-minute timeout
        
        # Initialize Apify SDK client
        self.apify_client = ApifyClient(self.api_token)
        self.initialized = False
        
    async def verify_connection(self) -> bool:
        """Verify API token and connection by fetching user data"""
        try:
            # Use direct HTTP request to get user info
            url = "https://api.apify.com/v2/users/me"
            headers = {"Authorization": f"Bearer {self.api_token}"}
            
            response = await self.client.get(url, headers=headers)
            response.raise_for_status()
            
            user_data = response.json()
            username = user_data.get("data", {}).get("username", "unknown")
            
            logger.info(f"Connected to Apify as user: {username}")
            self.initialized = True
            return True
        except Exception as e:
            logger.error(f"Failed to verify Apify connection: {e}")
            return False
    
    async def run_twitter_scraper_sync(self, handles: List[str], max_tweets: int = 10) -> List[Dict[str, Any]]:
        """Run Twitter scraper with synchronized API endpoint"""
        if not self.initialized and not await self.verify_connection():
            raise ConnectionError("Failed to connect to Apify API")
        
        logger.info(f"Starting synchronous Twitter scraper for {len(handles)} handles")
        
        try:
            # Prepare Twitter scraper input
            input_data = {
                "usernames": handles,
                "tweetsDesired": max_tweets,
                "includeReplies": False,
                "skipTweetReplyBlocks": True,
                "includeRetweets": False,
                "includeLinkToProfile": True,
                "waitBetweenRequests": 5
            }
            
            # Build direct synchronous API URL
            base_url = "https://api.apify.com/v2/acts/apidojo~twitter-scraper-lite/run-sync-get-dataset-items"
            
            # Make synchronous API call
            response = await self.client.post(
                base_url,
                params={"token": self.api_token},
                json=input_data,
                timeout=300.0  # 5-minute timeout for sync request
            )
            
            response.raise_for_status()
            tweets = response.json()
            
            logger.info(f"Retrieved {len(tweets)} tweets from Twitter scraper")
            return tweets
        except Exception as e:
            logger.error(f"Error running Twitter scraper: {e}")
            return []
    
    async def run_website_scraper_sync(self, urls: List[str], max_depth: int = 1) -> List[Dict[str, Any]]:
        """Run website scraper with synchronized API endpoint"""
        if not self.initialized and not await self.verify_connection():
            raise ConnectionError("Failed to connect to Apify API")
        
        logger.info(f"Starting synchronous website scraper for {len(urls)} URLs")
        
        try:
            # Prepare website scraper input
            input_data = {
                "startUrls": [{"url": url} for url in urls],
                "maxCrawlDepth": max_depth,
                "maxCrawlPages": 10 * len(urls),  # Limit to 10 pages per URL
                "maxPagesPerCrawl": 10 * len(urls),
                "limitCrawlDepth": True,
                "saveContent": True,
                "saveContentAsMarkdown": True,
                "blockRequests": "font, media, image, stylesheet, script",
                "headless": True
            }
            
            # Build direct synchronous API URL
            base_url = "https://api.apify.com/v2/acts/apify~website-content-crawler/run-sync-get-dataset-items"
            
            # Make synchronous API call
            response = await self.client.post(
                base_url,
                params={"token": self.api_token},
                json=input_data,
                timeout=600.0  # 10-minute timeout for sync request
            )
            
            response.raise_for_status()
            content_items = response.json()
            
            logger.info(f"Retrieved {len(content_items)} content items from website scraper")
            return content_items
        except Exception as e:
            logger.error(f"Error running website scraper: {e}")
            return []
    
    async def close(self):
        """Close the HTTP client"""
        if self.client:
            await self.client.aclose()
            logger.info("Apify connection closed")

File: src/connections/apify_connection.ts
import axios from ;
import { db } from '../server/db';
import { insertTwitterScrapeDataSchema, insertAddressCacheSchema } from '../shared/schema';
import { eq } from 'drizzle-orm';
import { twitterScrapeData, addressCache } from '../shared/schema';
import { logger } from '../utils/logger';

interface Tweet {
  username: string;
  tweet_id: string;
  content: string;
  timestamp: string;
  metadata?: any;
}

export class ApifyConnection {
  private readonly apiToken: string;
  private readonly endpoint: string;
  private static instance: ApifyConnection;

  private constructor() {
    this.apiToken = process.env.APIFY_API_TOKEN || '';
    this.endpoint = 'https://api.apify.com/v2/acts/apidojo~tweet-scraper/run-sync-get-dataset-items';
  }

  public static getInstance(): ApifyConnection {
    if (!ApifyConnection.instance) {
      ApifyConnection.instance = new ApifyConnection();
    }
    return ApifyConnection.instance;
  }

  private extractContractAddresses(content: string): string[] {
    // Match Ethereum addresses (0x followed by 40 hex characters)
    const addressRegex = /0x[a-fA-F0-9]{40}/g;
    return Array.from(new Set(content.match(addressRegex) || []));
  }

  private async updateAddressCache(address: string, tweetId: string): Promise<void> {
    try {
      const existingAddress = await db.query.addressCache.findFirst({
        where: eq(addressCache.address, address)
      });

      if (existingAddress) {
        await db
          .update(addressCache)
          .set({ lastSeen: new Date() })
          .where(eq(addressCache.address, address));
      } else {
        await db.insert(addressCache).values({
          address,
          source: 'twitter',
          sourceId: tweetId,
          metadata: { initialTweetId: tweetId }
        });
      }
    } catch (error) {
      logger.error(`Error updating address cache: ${error}`);
    }
  }

  public async fetchAndStoreTweets(): Promise<void> {
    try {
      const response = await axios.get(`${this.endpoint}?token=${this.apiToken}`);
      const tweets: Tweet[] = response.data;

      for (const tweet of tweets) {
        try {
          const contractAddresses = this.extractContractAddresses(tweet.content);
          
          // Store tweet data
          await db.insert(twitterScrapeData).values({
            username: tweet.username,
            tweetId: tweet.tweet_id,
            content: tweet.content,
            contractAddresses,
            timestamp: new Date(tweet.timestamp),
            metadata: tweet.metadata
          });

          // Update address cache for each contract address found
          for (const address of contractAddresses) {
            await this.updateAddressCache(address, tweet.tweet_id);
          }
        } catch (error) {
          logger.error(`Error processing tweet ${tweet.tweet_id}: ${error}`);
        }
      }
    } catch (error) {
      logger.error(`Error fetching tweets from Apify: ${error}`);
      throw error;
    }
  }
}

// Setup scheduled task (to be called from main application)
export const setupTwitterScraping = (): void => {
  const TWO_HOURS = 2 * 60 * 60 * 1000;
  const apifyConnection = ApifyConnection.getInstance();

  setInterval(async () => {
    try {
      await apifyConnection.fetchAndStoreTweets();
      logger.info('Successfully completed Twitter scraping cycle');
    } catch (error) {
      logger.error(`Error in Twitter scraping cycle: ${error}`);
    }
  }, TWO_HOURS);

  // Run initial scraping
  apifyConnection.fetchAndStoreTweets()
    .catch(error => logger.error(`Error in initial Twitter scraping: ${error}`));
};
